8주차 API 문서 만들기 & Docker Image 만들기

euijeonghwang opened the pull request:
마지막 주네요!
마지막까지 잘 부탁 드립니다!

euijeonghwang committed:
Feat: REST Docs를 사용한 API 문서 만들기 위한 의존성 추가

REST Docs를 위한 라이브러리 추가함

euijeonghwang committed:
Feat: REST Api를 좀 더 예쁘게 출력하기 위한 유틸 클래스 추가

우아한 형제들 기술블로그를 보고 따라서 설정함

euijeonghwang committed:
Test: Product Controller REST API 추가

ProductController 테스트 클래스를 실행할 때마다 API 생성함

euijeonghwang committed:
Feat: Product Rest Controller에 대한 API 문서 생성

기본적인 GET, POST, PATCH, DELETE에 대한 API 문서 작성함

euijeonghwang committed:
Style: AuthenticationService API 주석 작성

AuthenticationService API 문서를 위한 주석 작성함

euijeonghwang reviewed (commented):
⤵️결과물!
![Screenshot 2022-12-27 at 9 22 47 PM](https://user-images.githubusercontent.com/69239364/209666210-d52aadb0-ed74-4a4a-a6e1-85628c317440.png)

euijeonghwang commented:
API 문서는 여기까지 하고, Docker에 집중하려고 합니다!
물론 피드백은 반영하겠습니다!

euijeonghwang committed:
Feat: API 문서 html로 변경

결과물을 보기 위한 html 추가함

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -29,6 +69,19 @@ public AuthenticationService(UserRepository userRepository,
     }
 
+    /**
+     * Returns jwt of the user's id.
```
사용자의 id가 포함되어있는 JWT를 반환하는 것을 작성해 주셨네요. 이 메서드가 어떤 일을 하는지를 더 설명해주면 좋을 것 같습니다. 지금은 이 메서드의 동작을 설명하고 있는 것 같아요

hannut91 commented:
마지막까지 파이팅입니다!

euijeonghwang committed:
Style: AuthenticationService 주석 설명문 보충

조금 더 자세하게 설명 보충함

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -71,6 +71,11 @@ public AuthenticationService(UserRepository userRepository,
+     * If no such user is found, it throws a {@code LoginFailException}.
+     * If the user is found, it checks if the input password matches the password in the database.
+     * If it does not, it also throws a {@code LoginFailException}.
+     * If both the email and password are correct, it creates a JSON Web Token with the user's id as the payload and returns the token.
```
아주 자세히 써주셔서 좋네요 하지만 기능이 조금이라도 바뀌면 주석도 변경되어야 합니다. 기능이 변경되었는데 주석이 변경되지 않을 경우 대참사가 일어납니다.

어떻게 주석을 달면 동작이 변경되도 주석은 변경되지 않아도 될까요?

euijeonghwang replied:
음.. 145e17c에서 이메일을 받고 비밀번호를 비교한다는 내용 대신 입력된 정보를 가지고 유저를 찾는다라고 뭉뚱그려 표현해봤습니다!

euijeonghwang committed:
Feat: Docker image 만들기

Docker image 만들고, Dockerhub에 배포함
Dockerhub: https://hub.docker.com/r/ejhwang/cat-toy-server

euijeonghwang committed:
Feat: mariadb 라이브러리 추가

mariadb 사용을 위한 라이브러리 추가함

euijeonghwang committed:
Feat: mariadb 사용을 위한 프로파일 추가함

mariadb 사용을 위한 프로파일 추가함
3306포트 사용 중이어서 3305 사용함

euijeonghwang committed:
Style: 주석 추상화함

이메일, 비밀번호와 같이 구체적인 필드 대신 information이라는 추상적인 개념을 사용하여 메소드 설명함

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -70,12 +70,10 @@ public AuthenticationService(UserRepository userRepository,
     /**
-     * Returns jwt of the user's id.
-     * First, find the user with the given email in the database.
+     * Returns JSON Web Token of the user's id.
+     * Find the user with the given input user information in the database.
+     * When the user is found, it creates a JSON Web Token with the user's id as the payload and returns the token.
```
주석을 추상화하셨네요. 사용자를 찾는 방법이 달라져도 여기 주석이 달라지진 않겠네요. 

`JSON Web Token with the user's id` 이 부분이 조금 맘에걸리네요. JWT에 사용자 id뿐만 아니라 추가적인 정보를 넣고 싶을 수가 있고, 그러면 이 주석도 같이 변경되어야 겠네요.

hannut91 commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+# 위의 빌드 이미지로부터 만들어진 jar 파일을 복사합니다.
+COPY --from=builder /app/build/libs/app.jar .
+# 서버를 실행시키는 명령어입니다.
+CMD ["java", "-jar", "app.jar"]
```
여기는 ENTRYPOINT를 쓰는게 더 좋을거예요. 문서를 최신화 해야겠군요.

See also
  - https://www.bmc.com/blogs/docker-cmd-vs-entrypoint/

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -12,13 +12,53 @@
+     */
     private final JwtUtil jwtUtil;
+
+    /**
+     * Service interface for encoding passwords. It implements BCryptPasswordEncoder.
+     *
+     * @see org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
+     */
     private final PasswordEncoder passwordEncoder;
```
이 코드만 봐서는 BCryptPasswordEncoder를 구현 한다는 사실을 알기 어려울 것 같아요. 오히려 BCryptPasswordEncoder를 몰라야 하는게 맞는 것 같습니다. 우리는 언제든지 Encoder를 변경할 여지가 있습니다. 특히 BCrypt는 최근에는 사용하지 말라는 권고도 있었어요.

euijeonghwang replied:
BCryptPasswordEncoder 보다는 주로 SHA256 이상을 사용하나요?

hannut91 commented on app/src/main/resources/application.yml:
```diff
@@ -7,3 +7,13 @@ spring:
+spring:
+  profiles: mariadb
+  datasource:
+    url: jdbc:mariadb://mariadb:3305/test
```
Mariadb는 관습적으로 3306을 사용하는데 변경하셨군요

euijeonghwang replied:
3306 포트를 kill해도 계속해서 사용한다고 나와서 3305로 변경했습니다.

hannut91 replied:
아마도 데이터베이스가 동작하고 있나보군요.

euijeonghwang committed:
Style: 주석추상화

jwt를 유저의 id로 생성한다에서 jwt를 리턴한다로 변경함

euijeonghwang committed:
Feat: Dockerfile ENTRYPOTIN 사용

CMD 대신 ENTRYPOINT로 변경함

euijeonghwang committed:
Style: BCrpytPassword 에 관한 주석 제거

특정 암호화 대신 비밀번호 암호화를 도와주는 인터페이스라고 변경함

hannut91 reviewed (commented):
RESTDocs도 결국 클라이언트가 볼 수 있어야 하는데, 이 페이지는 어떻게 배포해야 할까요
